File: problems-extend-adapt/0601-0700/0676_one-edit-dictionary/solution.py
from typing import Dict, List


class OneEditDictionary:
    """Words grouped by length; each loadWords REPLACES the previous
    dictionary, so matchesOneEdit only ever sees the latest call's words.

    `matchesOneEdit` compares the matchesOneEdit word against every stored word of the
    same length and counts differing positions: a different length is
    simply no match, and true requires exactly one difference.
    """

    def __init__(self) -> None:
        self.buckets: Dict[int, List[str]] = {}

    def loadWords(self, dictionary: List[str]) -> None:
        buckets: Dict[int, List[str]] = {}
        for word in dictionary:
            buckets.setdefault(len(word), []).append(word)
        self.buckets = buckets

    def matchesOneEdit(self, searchWord: str) -> bool:
        for word in self.buckets.get(len(searchWord), []):
            mismatches = 0
            for candidate, wanted in zip(word, searchWord):
                if candidate != wanted:
                    mismatches += 1
                    if mismatches > 1:
                        break
            if mismatches == 1:
                return True
        return False

File: problems-extend-adapt/0601-0700/0676_one-edit-dictionary/solution.go
package main

// Words grouped by length; each loadWords REPLACES the previous dictionary,
// so matchesOneEdit only ever sees the latest call's words. A candidate matches when
// it differs from the matchesOneEdit word in exactly one position.

type OneEditDictionary struct {
	buckets map[int][]string
}

func NewOneEditDictionaryTyped() *OneEditDictionary {
	return &OneEditDictionary{buckets: map[int][]string{}}
}

func (design *OneEditDictionary) loadWords(dictionary []string) {
	buckets := map[int][]string{}
	for _, word := range dictionary {
		length := len(word)
		buckets[length] = append(buckets[length], word)
	}
	design.buckets = buckets
}

func (design *OneEditDictionary) matchesOneEdit(searchWord string) bool {
	for _, word := range design.buckets[len(searchWord)] {
		mismatches := 0
		for index := 0; index < len(word); index++ {
			if word[index] != searchWord[index] {
				mismatches++
				if mismatches > 1 {
					break
				}
			}
		}
		if mismatches == 1 {
			return true
		}
	}
	return false
}
